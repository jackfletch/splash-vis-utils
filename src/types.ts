export interface Point {
  x: number;
  y: number;
}

export interface Shot {
  id: number;
  game_id: number;
  game_event_id: number | string;
  player_id: number;
  distance: number;
  x: number;
  y: number;
  made_flag: boolean;
}
