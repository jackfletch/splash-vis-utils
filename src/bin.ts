import {isLessThanDistance} from "./filters";
import {LeftRight, Shot} from "./types";

export interface Bin {
  made: number;
  total: number;
}

function bin(shots: Shot[], maxDistance: number): Bin[] {
  const bins: Bin[] = Array(maxDistance + 1)
    .fill(undefined)
    .map(() => ({made: 0, total: 0}));

  shots.filter(isLessThanDistance(maxDistance)).forEach(shot => {
    const {distance, made} = shot;
    bins[distance].total += 1;
    bins[distance].made += +made;
  });

  return bins;
}

export function binLeftRight(
  shots: Shot[],
  maxDistance: number
): LeftRight<Bin[]> {
  const binsLeft: Bin[] = Array(maxDistance + 1)
    .fill(undefined)
    .map(() => ({made: 0, total: 0}));
  const binsRight: Bin[] = Array(maxDistance + 1)
    .fill(undefined)
    .map(() => ({made: 0, total: 0}));

  shots.filter(isLessThanDistance(maxDistance)).forEach(shot => {
    const {distance, made, x} = shot;
    if (x < 0) {
      binsLeft[distance].total += 1;
      binsLeft[distance].made += +made;
    } else if (x > 0) {
      binsRight[distance].total += 1;
      binsRight[distance].made += +made;
    }
  });

  return {left: binsLeft, right: binsRight};
}

export default bin;
