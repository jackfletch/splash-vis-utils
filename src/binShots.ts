import {Point, Shot} from "./types";

interface ShotDistanceMap {
  [index: number]: Shot[];
}

interface BinnedShots {
  binData: Point[];
  totalShots: number;
  totalMakes: number;
}

function binShots(data: Shot[], maxDistance: number): BinnedShots {
  const binnedShots: ShotDistanceMap = {};
  let totalMakes = 0;
  let totalShots = data.length;

  data.forEach(shot => {
    if (shot.distance <= maxDistance) {
      if (shot.distance in binnedShots) {
        binnedShots[shot.distance].push(shot);
      } else {
        binnedShots[shot.distance] = [shot];
      }
      if (shot.made_flag) {
        totalMakes += 1;
      }
    } else {
      totalShots -= 1;
    }
  });

  const binData: Point[] = [];
  let x = 0;

  Object.keys(binnedShots)
    .map(Number)
    .forEach(bin => {
      while (x !== bin) {
        binData.push({x, y: 0});
        x += 1;
      }
      const shots = binnedShots[bin];
      binData.push({x: bin, y: shots.length});
      x += 1;
    });

  const binnedData = {
    binData,
    totalShots,
    totalMakes,
  };
  return binnedData;
}

export default binShots;
