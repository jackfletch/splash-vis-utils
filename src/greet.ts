export interface Person {
  firstName?: string;
  lastName: string;
}

function greet(person: Person): void {
  console.log("hello " + person?.firstName);
}

export default greet;
