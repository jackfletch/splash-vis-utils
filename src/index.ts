export {default as bin, binLeftRight} from "./bin";
export {default as binShots} from "./binShots";
export {default as ribbonShots} from "./ribbonShots";
export {Point, Shot} from "./types";
