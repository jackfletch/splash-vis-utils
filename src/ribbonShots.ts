import binByDistance from "./bin";
import {Shot} from "./types";

interface RibbonBin {
  shootingPct: number;
  width: number;
}

function ribbonShots(data: Shot[], maxDistance: number): RibbonBin[] {
  const bins = binByDistance(data, maxDistance);
  const totalShots = bins.reduce((a, b) => a + b.total, 0);

  const ribbonBins = bins.map((bin) => {
    const shootingPct = bin.total > 0 ? bin.made / bin.total : 0;
    const width = bin.total / totalShots;
    return {
      shootingPct,
      width,
    };
  });

  // TODO: implement smoothing function

  return ribbonBins;
}

export default ribbonShots;
