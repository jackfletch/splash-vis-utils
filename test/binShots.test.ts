import {binShots} from "../src";
import {DEFAULT_DISTANCE, MAX_DISTANCE, SMALL_DISTANCE} from "./constants";
import shot2544 from "./fixtures/shots/2544.json";

describe("binShots", () => {
  it(`Should correctly bin shots with max distance of ${DEFAULT_DISTANCE}`, () => {
    const result = binShots(shot2544, DEFAULT_DISTANCE);
    expect(result).toMatchSnapshot();
  });

  it(`Should correctly bin shots with max distance of ${MAX_DISTANCE}`, () => {
    const result = binShots(shot2544, MAX_DISTANCE);
    expect(result).toMatchSnapshot();
  });

  it(`Should correctly bin shots with max distance of ${SMALL_DISTANCE}`, () => {
    const result = binShots(shot2544, SMALL_DISTANCE);
    expect(result).toMatchSnapshot();
  });

  it("Should correctly bin shots given only one shot", () => {
    const shots = [
      {
        id: 100442,
        game_id: 21600001,
        game_event_id: "8",
        player_id: 2544,
        distance: 11,
        x: -79,
        y: 80,
        made: true,
      },
    ];
    const result = binShots(shots, DEFAULT_DISTANCE);
    expect(result).toMatchSnapshot();
  });
});
