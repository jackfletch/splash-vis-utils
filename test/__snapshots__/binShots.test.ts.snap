// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`binShots Should correctly bin shots given only one shot: default 1`] = `
Object {
  "binData": Array [
    Object {
      "x": 0,
      "y": 0,
    },
    Object {
      "x": 1,
      "y": 0,
    },
    Object {
      "x": 2,
      "y": 0,
    },
    Object {
      "x": 3,
      "y": 0,
    },
    Object {
      "x": 4,
      "y": 0,
    },
    Object {
      "x": 5,
      "y": 0,
    },
    Object {
      "x": 6,
      "y": 0,
    },
    Object {
      "x": 7,
      "y": 0,
    },
    Object {
      "x": 8,
      "y": 0,
    },
    Object {
      "x": 9,
      "y": 0,
    },
    Object {
      "x": 10,
      "y": 0,
    },
    Object {
      "x": 11,
      "y": 1,
    },
  ],
  "totalMakes": 1,
  "totalShots": 1,
}
`;

exports[`binShots Should correctly bin shots with max distance of 5: 5 1`] = `
Object {
  "binData": Array [
    Object {
      "x": 0,
      "y": 316,
    },
    Object {
      "x": 1,
      "y": 180,
    },
    Object {
      "x": 2,
      "y": 87,
    },
    Object {
      "x": 3,
      "y": 32,
    },
    Object {
      "x": 4,
      "y": 18,
    },
    Object {
      "x": 5,
      "y": 25,
    },
  ],
  "totalMakes": 485,
  "totalShots": 658,
}
`;

exports[`binShots Should correctly bin shots with max distance of Number.MAX_SAFE_INTEGRER: max 1`] = `
Object {
  "binData": Array [
    Object {
      "x": 0,
      "y": 316,
    },
    Object {
      "x": 1,
      "y": 180,
    },
    Object {
      "x": 2,
      "y": 87,
    },
    Object {
      "x": 3,
      "y": 32,
    },
    Object {
      "x": 4,
      "y": 18,
    },
    Object {
      "x": 5,
      "y": 25,
    },
    Object {
      "x": 6,
      "y": 18,
    },
    Object {
      "x": 7,
      "y": 19,
    },
    Object {
      "x": 8,
      "y": 14,
    },
    Object {
      "x": 9,
      "y": 15,
    },
    Object {
      "x": 10,
      "y": 14,
    },
    Object {
      "x": 11,
      "y": 19,
    },
    Object {
      "x": 12,
      "y": 20,
    },
    Object {
      "x": 13,
      "y": 30,
    },
    Object {
      "x": 14,
      "y": 26,
    },
    Object {
      "x": 15,
      "y": 32,
    },
    Object {
      "x": 16,
      "y": 28,
    },
    Object {
      "x": 17,
      "y": 18,
    },
    Object {
      "x": 18,
      "y": 14,
    },
    Object {
      "x": 19,
      "y": 23,
    },
    Object {
      "x": 20,
      "y": 20,
    },
    Object {
      "x": 21,
      "y": 23,
    },
    Object {
      "x": 22,
      "y": 20,
    },
    Object {
      "x": 23,
      "y": 23,
    },
    Object {
      "x": 24,
      "y": 74,
    },
    Object {
      "x": 25,
      "y": 119,
    },
    Object {
      "x": 26,
      "y": 62,
    },
    Object {
      "x": 27,
      "y": 19,
    },
    Object {
      "x": 28,
      "y": 14,
    },
    Object {
      "x": 29,
      "y": 9,
    },
    Object {
      "x": 30,
      "y": 4,
    },
    Object {
      "x": 31,
      "y": 2,
    },
    Object {
      "x": 32,
      "y": 3,
    },
    Object {
      "x": 33,
      "y": 2,
    },
    Object {
      "x": 34,
      "y": 0,
    },
    Object {
      "x": 35,
      "y": 0,
    },
    Object {
      "x": 36,
      "y": 0,
    },
    Object {
      "x": 37,
      "y": 0,
    },
    Object {
      "x": 38,
      "y": 0,
    },
    Object {
      "x": 39,
      "y": 0,
    },
    Object {
      "x": 40,
      "y": 1,
    },
  ],
  "totalMakes": 736,
  "totalShots": 1343,
}
`;

exports[`binShots Should correctly bin shots: default 1`] = `
Object {
  "binData": Array [
    Object {
      "x": 0,
      "y": 316,
    },
    Object {
      "x": 1,
      "y": 180,
    },
    Object {
      "x": 2,
      "y": 87,
    },
    Object {
      "x": 3,
      "y": 32,
    },
    Object {
      "x": 4,
      "y": 18,
    },
    Object {
      "x": 5,
      "y": 25,
    },
    Object {
      "x": 6,
      "y": 18,
    },
    Object {
      "x": 7,
      "y": 19,
    },
    Object {
      "x": 8,
      "y": 14,
    },
    Object {
      "x": 9,
      "y": 15,
    },
    Object {
      "x": 10,
      "y": 14,
    },
    Object {
      "x": 11,
      "y": 19,
    },
    Object {
      "x": 12,
      "y": 20,
    },
    Object {
      "x": 13,
      "y": 30,
    },
    Object {
      "x": 14,
      "y": 26,
    },
    Object {
      "x": 15,
      "y": 32,
    },
    Object {
      "x": 16,
      "y": 28,
    },
    Object {
      "x": 17,
      "y": 18,
    },
    Object {
      "x": 18,
      "y": 14,
    },
    Object {
      "x": 19,
      "y": 23,
    },
    Object {
      "x": 20,
      "y": 20,
    },
    Object {
      "x": 21,
      "y": 23,
    },
    Object {
      "x": 22,
      "y": 20,
    },
    Object {
      "x": 23,
      "y": 23,
    },
    Object {
      "x": 24,
      "y": 74,
    },
    Object {
      "x": 25,
      "y": 119,
    },
    Object {
      "x": 26,
      "y": 62,
    },
    Object {
      "x": 27,
      "y": 19,
    },
    Object {
      "x": 28,
      "y": 14,
    },
    Object {
      "x": 29,
      "y": 9,
    },
    Object {
      "x": 30,
      "y": 4,
    },
    Object {
      "x": 31,
      "y": 2,
    },
    Object {
      "x": 32,
      "y": 3,
    },
    Object {
      "x": 33,
      "y": 2,
    },
  ],
  "totalMakes": 736,
  "totalShots": 1342,
}
`;
